src/util/websocket/WebSocketManager.cpp

```diff
@@ -0,0 +1,177 @@
+
+#include "WebSocketManager.h"
+
+#include <ctre/ctre.h>
+
+#include <boost/asio/experimental/awaitable_operators.hpp>
+#include <optional>
+
+#include "util/http/HttpUtils.h"
+#include "util/websocket/Common.h"
+
+namespace ad_utility::websocket {
+using namespace boost::asio::experimental::awaitable_operators;
+
+void WebSocketManager::addQueryStatusUpdate(const QueryId& queryId,
+                                            std::string payload) {
+  auto sharedPayload = std::make_shared<const std::string>(std::move(payload));
+  net::post(registryStrand_.value(),
+            [this, sharedPayload = std::move(sharedPayload), &queryId]() {
+              if (informationQueues_.contains(queryId)) {
+                informationQueues_.at(queryId).append(std::move(sharedPayload));
+              } else {
+                informationQueues_.emplace(
+                    queryId,
+                    ad_utility::LinkedList<std::shared_ptr<const std::string>>{
+                        std::move(sharedPayload)});
+              }
+              wakeUpWebSocketsForQuery(queryId);
+            });
+}
+
+void WebSocketManager::wakeUpWebSocketsForQuery(const QueryId& queryId) {
+  net::post(registryStrand_.value(), [this, &queryId]() {
+    auto range = wakeupCalls_.equal_range(queryId);
+    for (auto it = range.first; it != range.second; ++it) {
+      it->second();
+    }
+    wakeupCalls_.erase(queryId);
+  });
+}
+
+void WebSocketManager::releaseQuery(QueryId queryId) {
+  net::post(registryStrand_.value(), [this, queryId = std::move(queryId)]() {
+    informationQueues_.erase(queryId);
+    auto range = wakeupCalls_.equal_range(queryId);
+    for (auto it = range.first; it != range.second; ++it) {
+      it->second();
+    }
+    wakeupCalls_.erase(queryId);
+  });
+}
+
+// Based on https://stackoverflow.com/a/69285751
+template <typename CompletionToken>
+auto UpdateFetcher::waitForEvent(CompletionToken&& token) {
+  auto initiate = [this]<typename Handler>(Handler&& self) mutable {
+    auto sharedSelf = std::make_shared<Handler>(std::forward<Handler>(self));
+
+    if (currentNode_ && currentNode_->hasNext()) {
+      currentNode_ = currentNode_->next();
+      (*sharedSelf)(currentNode_->getPayload());
+    } else {
+      net::post(registryStrand_, [this, sharedSelf]() {
+        wakeupCalls_.insert(
+            {queryId_, [this, sharedSelf]() {
+               if (currentNode_) {
+                 if (!currentNode_->hasNext()) {
+                   (*sharedSelf)(nullptr);
+                 } else {
+                   currentNode_ = currentNode_->next();
+                   (*sharedSelf)(currentNode_->getPayload());
+                 }
+               } else {
+                 if (informationQueues_.contains(queryId_)) {
+                   currentNode_ = informationQueues_.at(queryId_).getHead();
+                   (*sharedSelf)(currentNode_->getPayload());
+                 } else {
+                   (*sharedSelf)(nullptr);
```

Comment:
This is where I'm really unsure about the behaviour of boost asio:
When calling `sharedSelf` here, does this mean the co_await will resume on the `registryStrand_` or on the initial strand it was invoked on? Not sure how to check this, but the former behaviour would be pretty bad in terms of performance